File: carfo-backend/src/main/java/com/carfo/gestion_missions/service/AuditService.java
package com.carfo.gestion_missions.service;

import com.carfo.gestion_missions.entity.Agent;
import com.carfo.gestion_missions.entity.AuditLog;
import com.carfo.gestion_missions.enums.AuditCategory;
import com.carfo.gestion_missions.repository.AuditLogRepository;
import com.carfo.gestion_missions.security.SecurityChecker;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;

/**
 * Enregistre les actions sensibles dans la table audit_log.
 * Échoue silencieusement : un problème d'audit ne doit JAMAIS interrompre le flow métier.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class AuditService {

    private final AuditLogRepository repository;
    private final SecurityChecker securityChecker;

    /** Injection optionnelle de la requête en cours pour capturer l'IP (proxy-friendly). */
    @Autowired(required = false)
    private HttpServletRequest currentRequest;

    /**
     * Enregistre une entrée d'audit pour l'utilisateur courant (résolu via SecurityChecker).
     * Si aucun utilisateur n'est authentifié (cas LOGIN avant succès), les champs agent restent vides.
     */
    @Transactional
    public void log(AuditCategory category, String action, String entityType, Long entityId, String summary) {
        var currentAgent = securityChecker.getCurrentAgent().orElse(null);
        logAs(currentAgent, category, action, entityType, entityId, summary);
    }

    /**
     * Variante explicite : enregistre une entrée d'audit en imputant l'action à un agent spécifique
     * (utilisée notamment lors du LOGIN, où l'utilisateur n'est pas encore dans le SecurityContext).
     */
    @Transactional
    public void logAs(Agent agent, AuditCategory category, String action,
                       String entityType, Long entityId, String summary) {
        try {
            AuditLog entry = AuditLog.builder()
                    .category(category)
                    .action(action)
                    .entityType(entityType)
                    .entityId(entityId)
                    .summary(truncate(summary, 500))
                    .ipAddress(resolveClientIp())
                    .agentEmail(agent != null ? agent.getEmail() : null)
                    .agentNom(agent != null ? (agent.getPrenom() + " " + agent.getNom()) : null)
                    .build();
            repository.save(entry);
        } catch (Exception ex) {
            // Erreur d'audit non-bloquante (par ex. connexion DB momentanée).
            log.warn("Audit log skipped ({}/{}): {}", category, action, ex.getMessage());
        }
    }

    @Transactional(readOnly = true)
    public Page<AuditLog> search(AuditCategory category, String email,
                                  LocalDateTime fromDate, LocalDateTime toDate,
                                  int page, int size) {
        int safePage = Math.max(0, page);
        int safeSize = Math.min(200, Math.max(1, size));
        return repository.search(category, email, fromDate, toDate,
                PageRequest.of(safePage, safeSize, Sort.by(Sort.Direction.DESC, "timestamp")));
    }

    private String resolveClientIp() {
        if (currentRequest == null) return null;
        String forwarded = currentRequest.getHeader("X-Forwarded-For");
        if (forwarded != null && !forwarded.isBlank()) {
            return forwarded.split(",")[0].trim();
        }
        return currentRequest.getRemoteAddr();
    }

    private String truncate(String s, int max) {
        if (s == null) return null;
        return s.length() <= max ? s : s.substring(0, max);
    }
}

File: carfo-frontend/src/app/core/services/audit.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export type AuditCategory = 'AUTH' | 'MISSION' | 'AFFECTATION' | 'AGENT' | 'CONFIG' | 'OTHER';

export interface AuditLog {
  idAudit: number;
  timestamp: string;
  agentEmail?: string | null;
  agentNom?: string | null;
  category: AuditCategory;
  action: string;
  entityType?: string | null;
  entityId?: number | null;
  summary?: string | null;
  ipAddress?: string | null;
}

export interface AuditPage {
  content: AuditLog[];
  totalElements: number;
  totalPages: number;
  page: number;
  size: number;
}

export interface AuditFilters {
  category?: AuditCategory;
  email?: string;
  fromDate?: string; // ISO LocalDateTime
  toDate?: string;
  page?: number;
  size?: number;
}

@Injectable({ providedIn: 'root' })
export class AuditService {
  private readonly API_URL = `${environment.apiUrl}/audit`;

  constructor(private readonly http: HttpClient) {}

  search(filters: AuditFilters = {}): Observable<AuditPage> {
    let params = new HttpParams();
    if (filters.category) params = params.set('category', filters.category);
    if (filters.email?.trim()) params = params.set('email', filters.email.trim());
    if (filters.fromDate) params = params.set('fromDate', filters.fromDate);
    if (filters.toDate) params = params.set('toDate', filters.toDate);
    params = params.set('page', String(filters.page ?? 0));
    params = params.set('size', String(filters.size ?? 30));
    return this.http.get<AuditPage>(this.API_URL, { params });
  }
}
